drivers/flash/flash_npcx_fiu_qspi.c

```diff
@@ -0,0 +1,285 @@
+#define NPCX_UMA_TIMEOUT 10
+
+/* QSPI bus mode for Direct Read Access */
+enum npcx_qspi_fiu_read_mode {
+	 /* SPI NOR Normal read */
+	 NPCX_QSPI_FIU_READ_NOR = 0,
+	 /* SPI NOR Fast read */
+	 NPCX_QSPI_FIU_READ_FAST,
+	 /* Reserved */
+	 NPCX_QSPI_FIU_READ_RESV,
+	 /* SPI NOR Fast Read Dual I/O */
+	 NPCX_QSPI_FIU_READ_FAST_DUAL,
+};
+
+/* Device config */
+struct npcx_qspi_fiu_config {
+	/* Flash interface unit base address */
+	uintptr_t base;
+	/* Clock configuration */
+	struct npcx_clk_cfg clk_cfg;
+	/* Enable 2 external SPI devices for direct read on QSPI bus */
+	bool en_direct_access_2dev;
+};
+
+/* Device data */
+struct npcx_qspi_fiu_data {
+	/* mutex of qspi bus controller */
+	struct k_sem lock_sem;
+	/* Active device on QSPI bus */
+	const struct npcx_qspi_cfg *active;
+	/* Current Software controlled Chip-Select number */
+	int sw_cs;
+	/* Current QSPI bus operation */
+	uint32_t operation;
+};
+
+/* NPCX SPI User Mode Access (UMA) functions */
+static inline void qspi_npcx_uma_cs_level(const struct device *dev, int sw_cs, int level)
+{
+	struct fiu_reg *const inst = HAL_INSTANCE(dev);
+
+	/* Set chip select to high/low level */
+	if (level == 0)
+		inst->UMA_ECTS &= ~BIT(sw_cs);
+	else
+		inst->UMA_ECTS |= BIT(sw_cs);
+}
+
+static inline void qspi_npcx_uma_write_byte(const struct device *dev, uint8_t data)
+{
+	struct fiu_reg *const inst = HAL_INSTANCE(dev);
+
+	/* Set data to UMA_CODE and trigger UMA */
+	inst->UMA_CODE = data;
+	inst->UMA_CTS = UMA_CODE_CMD_WR_ONLY;
+	/* EXEC_DONE will be zero automatically if a UMA transaction is completed. */
+	while (IS_BIT_SET(inst->UMA_CTS, NPCX_UMA_CTS_EXEC_DONE)) {
+		continue;
+	}
+}
+
+static inline void qspi_npcx_uma_read_byte(const struct device *dev, uint8_t *data)
+{
+	struct fiu_reg *const inst = HAL_INSTANCE(dev);
+
+	/* Trigger UMA and Get data from DB0 later */
+	inst->UMA_CTS = UMA_CODE_RD_BYTE(1);
+	while (IS_BIT_SET(inst->UMA_CTS, NPCX_UMA_CTS_EXEC_DONE)) {
+		continue;
+	}
+
+	*data = inst->UMA_DB0;
+}
+
+/* NPCX SPI Direct Read Access (DRA)/User Mode Access (UMA) configuration functions */
+static inline void qspi_npcx_config_uma_mode(const struct device *dev,
+					     const struct npcx_qspi_cfg *qspi_cfg)
+{
+	struct fiu_reg *const inst = HAL_INSTANCE(dev);
+
+	if ((qspi_cfg->flags & NPCX_QSPI_SEC_FLASH_SL) != 0) {
+		inst->UMA_ECTS |= BIT(NPCX_UMA_ECTS_SEC_CS);
+	} else {
+		inst->UMA_ECTS &= ~BIT(NPCX_UMA_ECTS_SEC_CS);
+	}
+}
+
+static inline void qspi_npcx_config_dra_mode(const struct device *dev,
+					     const struct npcx_qspi_cfg *qspi_cfg)
+{
+	struct fiu_reg *const inst = HAL_INSTANCE(dev);
+
+	/* Enable quad mode in Direct rRead Mode? */
+	if (qspi_cfg->qer_type != JESD216_DW15_QER_NONE) {
+		/* Set Read Mode to Fast Read Dual I/O and enable quad mode */
+		SET_FIELD(inst->SPI_FL_CFG, NPCX_SPI_FL_CFG_RD_MODE,
+					NPCX_QSPI_FIU_READ_FAST_DUAL);
+		inst->RESP_CFG |= BIT(NPCX_RESP_CFG_QUAD_EN);
+	} else {
+		inst->RESP_CFG &= ~BIT(NPCX_RESP_CFG_QUAD_EN);
```

Comment:
>Should this path set the RD_MODE field?

Usually, the driver chooses `Fast Read Dual I/O` mode by default for better performance since most of nor spi flashes support it. But we cannot expect all of them to support this access mode. I will add a dt-property `rd-mode` for the `RD_MODE` field. If the spi nor flash does not support `Fast Read Dual I/O` mode, the user can set this property explicitly. 

>Also, is fast dual I/O mode supported when using the write protection feature?

When DRA is configured to `Fast Dual I/O` mode, it only uses four pins of spi nor flash. It won't use the write protect pin.

Reply:
Done.